### ISSUE TITLE
How to perform unattended system upgrades?

### ISSUE THREAD

**Issue by jokogr** [NixOS/nixops#842]
In conventional NixOS deployments there is the `system.autoUpgrade` option which setups a systemd timer and service which perform periodically channel upgrades and `nixos-rebuild switch`.

Given that nixops does not deploy the configuration to `/etc/nixos` (so `system.autoUpgrade` cannot be effectively used), is there anything else that people are using for the same purpose?

**Comment by jazmit** [comment]
+1 There are blog posts suggesting to manually copy the configuration which seems to work, but I would appreciate some opinions from the maintainers on this..

**Comment by craigem** [comment]
The only blog post I've been able to find on this issue is:

[Auto-upgrade with NixOS and NixOps](https://pascal-wittmann.de/category/Blog/entry/autoupgrade-with-nixops-and-nixops) by [Pascal Wittmann](https://pascal-wittmann.de/).

**Comment by asymmetric** [comment]
Why doesn’t NixOps copy the `configuration.nix` to `/etc`?

**Comment by pSub** [comment]
@craigem Does my approach work for you? My configuration has changes slightly since the post. You can find the current version [here](https://github.com/pSub/configs/blob/master/nixops/configurations/server.pascal-wittmann.de/default.nix#L47). However my auto-upgrade is currently broken, because I use `deployment` in a module that is `require`d. The `filterAttrs` has no chance of sorting that out. By the way the issue NixOS/nixpkgs#28527 I mention in my config seems to be resolved in nixpkgs `master`.

**Comment by MelleB** [comment]
My hacky work-around is the following 2 lines in my `configuration.nix`:

```
environment.etc."nixos/configuration.nix".text = builtins.readFile "/full/path/to/configuration.nix";
environment.etc."nixos/hardware-configuration.nix".text = builtins.readFile "/full/path/to/hardware-configuration.nix";
```

Hope that helps someone.

**Comment by pSub** [comment]
@MelleB Do you copy `configuration.nix` manually to the server? Or how is full/path/to/configuration.nix realized? The nice thing about my approach is that no manual intervention is needed.

**Comment by asymmetric** [comment]
@pSub which specific parts of your config are you referring to?

**Comment by pSub** [comment]
@asymmetric Yes, but you need the activation script and nixos-upgrade-path too. [This](https://github.com/pSub/configs/blob/master/nixops/configurations/server.pascal-wittmann.de/default.nix#L58-L80) should be all you need besides the standard nixos-upgrade options.
This copies all configuration files (except the keys and passwords) into `/etc/nixos/current`. The `configuration.nix` uses the `server` attribute from the nixops deployment configuration and removes `deployment` attributes and `require`s that have `deployment` in their filename. This is needed because `nixos-rebuild` does not know about the `deployment` attribute.
I hope this makes sense.

**Comment by aanderse** [comment]
Maybe I'm missing something... but if you want automatic upgrades on `nixops` why wouldn't you just make a job that runs `nixops deploy` on the `nixops` master server? :man_shrugging:

**Comment by benley** [comment]
> Maybe I'm missing something... but if you want automatic upgrades on `nixops` why wouldn't you just make a job that runs `nixops deploy` on the `nixops` master server?

That doesn't really work if you run nixops from a laptop, or a workstation that's not always online.  It's not a workflow I'd recommend for Real World Business Use, but I suspect it's a common scenario for individual users and SOHO deployments.
